### ISSUE TITLE
Rails5, auto-loading, future, circular dependency

### ISSUE THREAD

**Issue by jrochkind** [ruby-concurrency/concurrent-ruby#585]
My first time trying to use concurrent-ruby in Rails5 and with MRI 2.3.  MRI 2.3.1. concurrent-ruby 1.0.2. Running into something very odd. 

    future = Concurrent::Future.execute { something_that_will_cause_rails_autoloading }
    future.value

Hangs forever, never comes back. It's only by some weird debugging that I discovered it's "something_that_will_cause_autoloading" -- and I'm not _certain_ of this, but I suspect it was Rails auto-loading behavior,   How?  Before calling future.value, I did a weird `future.instance_variable_get('@executor').shutdown`.  After I do that, if I call `future.value`, it does return, but has failed -- and `future.reason` is `#<RuntimeError: Circular dependency detected while autoloading constant BentoSearch::ResultItem>` (which is one of my own classes)

So some thoughts:

1. I am not sure why it's deciding there's a 'circular dependency', I can `require 'bento_search/result_item'` on it's own (not in a future) fine, with no warnings. 
2. I suspect I am doing something Rails doesn't like, perhaps you _need_ to use some kind of eager loading if you are going to use concurrency in Rails 5? 
3. But even if I'm doing something wrong, why is the future hanging forever?  What's actually going on?  I'm not sure. I'd expect it to come back with an exception -- perhaps that "circular dependency" exception, why did I end up with that if I reached in to find the executor and shut it down myself, but a hang forever otherwise?  It seems like concurrent-ruby is maybe doing something wrong.... maybe?

Wow, concurrency is confusing. Any thoughts.

**Comment by pitr-ch** [comment]
@jrochkind Thanks for letting us know! We definitely pay attention to rails + c-r combination. Could you create a reproducer for this? The blocking part is enough.

**Comment by jrochkind** [comment]
Upon further investigation, I think it is about Rails auto-loading, and is all about Rails, and not at a  ruby-concurrent bug or something ruby-concurrent can do anything about.

But is something users of ruby-concurrent in a Rails app will have to know about, that you may want to give them hints on or be able to answer questions on. 

Repro with giant write-up [here](https://github.com/jrochkind/rails5_concurrency_autoload_repro), also reported as [Rails issue](https://github.com/rails/rails/issues/26847). 

Basically, using concurrency in the request-loop in Rails5 seems to require turning off dev-mode class reloading, and the 'failure mode' if you fail to do so looks like a deadlock.

**Comment by jrochkind** [comment]
Okay, I've figured some stuff out by reading source, expirimentation, but mostly by @matthewd answering my questions. 

Rails 5 has several new architectural components and API for dealing with concurrency. 

The good news is that can make auto-loading, and even dev-mode reloading, concurrency-safe for the first time ever. 

The bad news it's pretty much entirely undocumented (not even method-level rdoc mostly), and if you don't use it correctly the 'failure mode' can be a deadlock in your application.  As opposed to Rails4, where autoload concurrency failure meant weird exceptions, now it can often mean deadlock. 

The 'correct' way to do some concurrency inside a Rails action method and then wait for the work to complete before returning a response:

~~~ruby
  # a controller action method
  def show
    futures = 3.times.collect do |i|
      Concurrent::Future.execute do
        Rails.application.executor.wrap do
          SomeWorker.new.value
       end
      end
    end

    ActiveSupport::Dependencies.interlock.permit_concurrent_loads do
      @example_values = futures.collect(&:value)
    end
  end
~~~

* ActiveSupport::Dependencies.interlock.permit_concurrent_loads

   In cases where autoloading happens, if you don't do this, you wind up with deadlock, where the threads are trying to autoload, but can't, because the Rails request/action/worker has some type of lock on autoloading. The future threads are waiting for the lock to be relinquished, but the worker thread with the lock is waiting for the threads to be complete. 

   So in those cases, the threads will actually be blocked until we do `permit_concurrent_loads`, at which point they'll be allowed to do their thing, and we wait for them to complete inside that block. 

   If autoloading isn't happening, this isn't necessary. So for instance, if Rails `config.eager_load == config.cache_classes == true` (as normally in production, but _not_ dev), then I don't _think_ it's neccesary. 

    It _probably_ doesn't hurt even when used where not necessary. 

* Rails.application.executor.wrap
    Rails new API for signaling intent 'to do some work'.  One thing it does is make autoloading actually work concurrency-safe.  If you don't have this line here, and autoloading happens inside the future (actually is triggered, not just an auto-loadable constant, but when which has not yet been loaded), then one or more futures may raise `RuntimeError: Circular dependency detected while autoloading constant NameOfClass`.  I am not sure why this exhibits with a message "Circular dependency" or what's "circular" about it, but it means "you didn't autoload in a concurrency safe way". 

    `Rails.application.executor.wrap` around your work is the way to make autoloading concurrency safe. But it's intended to be used regardless of dev or production mode, it's always intended to be used around 'Rails-related work being done' in a thread. 
   
    It _may_ eliminate the need for ye old `ActiveRecord::Base.connection_pool.with_connection`, the wrap takes care of checking connections back in at the end of the block for you. Needs more verification. 

    There's another version, `Rails.application.reloader.wrap`.  I believe the functionality there is a _superset_  of `Rails.application.executor.wrap`, but I'm not totally sure in what ways. But you can't use it in this case, since a Rails request loop is already wrapped in a `Rails.application.reloader.wrap`, using one here... seems to result in deadlock. [Sidekiq](https://github.com/mperham/sidekiq/pull/2457) does use it for it's workers, which I think result in the workers actually getting dev-mode reloading working for them. So that's one extra thing it does. 

Phew. I haven't fully wrapped my head around how this impacts using gems which use concurrency (with concurrent-ruby or otherwise) that aren't neccessarily written specifically for Rails, so don't have any Rails specific API in them and don't want any  -- if they end up triggering any Rails auto-loading, it could be bad and deadlock.  They probably wouldn't ordinarily, unless you pass class names into them or something. 

@matthewd promises some actual docs, which I eager await. If anyone else wants to play around with concurrent-ruby in a Rails app, your feedback or problems you run into might help @matthewd understand his audience and what the docs need to hit.

**Comment by jrochkind** [comment]
So, yeah, it's not a concurrent-ruby issue... but I think many many people trying to use concurrent-ruby in Rails5 are going to run into these issues, so concurrent-ruby team probably wants to be aware of them, and maybe even doc them especially if Rails doesn't doc them soon.

**Comment by ianks** [comment]
I stumbled across this issue as well. Spend an afternoon trying to figure out why I started deadlocking. I think dealing with this should be high-priority since it really highers the burden of entry on using concurrent patterns in Rails.

**Comment by pitr-ch** [comment]
yeah that's a good idea, I was thinking that we should have a bridge/adapter to make c-r work easily under Rails. However I am no longer actively working with Rails applications and I do not have time to do it. I'll support anyone (and help when I can) who would decide to take up the task. @ianks ?

**Comment by jrochkind** [comment]
I'm honestly not sure there's any good way to take care of it for the user, the Rails architecture may just need the user to express their intent as intended. Good docs supplementing Rails docs may be the best that can be done.

**Comment by pitr-ch** [comment]
I was hoping to wrap c-r tasks with permit_concurrent_loads or something like that. For promises it might be easy, theirs FactoryMethods were designed to support cases like that. Other abstractions might be problematic. I did not thing about it in depth at all. Even if the gem provides only few key abstractions which will work with Rails it might be a big win for users.

**Comment by matthewd** [comment]
I'm happy to provide information on what Rails needs etc.

I've given up on addressing it myself -- all my attempts have been vetoed because they involve providing APIs that a sufficiently-misguided library could misuse, and I still see no way around that.

**Comment by ianks** [comment]
It would be nice to have a blessed way of doing concurrency inside of the Rails request life cycle. I'm finding myself commonly wanting to resolve from external APIs, and I imagine it's a common use case. I've done alright for myself building abstractions which do the Rails.application.executor dance, but I think it's error-prone and tedious.

@matthewd, do you think a proposal for something like ActiveSupport::Concurrency::Promise would be considered? Or would this be something that should be third party?

**Comment by edwardmp** [comment]
@jrochkind 

Thanks for your write-up. It helped me tremendously when I encountered the 'circular dependency detected while autoloading constant NameOfClass.' issue.  I don't know much about threading in Ruby in general.

Here are my observations:

* Not using `Rails.application.executor.wrap` sometimes leads to the circular dependency issue. Actually in most cases it did not, but it still happened from time to time.
* After using `Rails.application.executor.wrap` my process went into a deadlock. 
* After calling `ActiveSupport::Dependencies.interlock.permit_concurrent_loads` at the end of the future execution resolves the deadlock. And yes, `config.cache_classes` is set to true (production) but I do actually need this (you thought you wouldn't need it with `config.cache_classes` enabled).

With regard to the last issue, if I look at your example the `permit_concurrent_loads` is not called from within the future execution cycle.  So although this works, I am wondering if this makes sense?

Here's a bit of sample code:
```
@beanstalk.jobs.register(tube_name) do |job|
    @logger.info "Received job: #{job.id}"
    Concurrent::Future.execute(executor: pool) do
      Rails.application.executor.wrap do
        process_beanstalk_message(job.body)
      end

      ActiveSupport::Dependencies.interlock.permit_concurrent_loads
    end
end
```

**Comment by jrochkind** [comment]
@edwardmp that's odd that you need this apparatus in production, and disturbing to me. I don't understand what's going on. Just to check, you say you have `cache_classes=true`, do you also have `config.eager_load = true` (as would be standard/default in production)?  @matthewd, any ideas or explanation of why you'd need to use the executor and interlock objects to do concurrency even in production configurations?

**Comment by jrochkind** [comment]
So a naive-implementation of a Rails-safe Future, for instance, might just be one that always wraps it's body in `Rails.application.executor.wrap`, and wraps `value` in `ActiveSupport::Dependencies.interlock.permit_concurrent_loads`. 

I'm not sure of the performance implications. 

If you are not in a situation that requires this (I _thought_ if you are in prod-mode configurations without autoloading it should not be needed, but may not be true according to reports above), are you paying a non-trivial expense?

How expensive is `ActiveSupport::Dependencies.interlock.permit_concurrent_loads` in cases where nothing is going to end up autoloaded anyway, are you paying a non-trivial cost by wrapping _each_ individual `value` in it?  

@matthewd can you give any tips or guidance on the performance considerations? 

One could imagine a rails-safe implementation of (eg) Future that actually feature-checked to make sure the relevant methods were available, so you could use it in any version of Rails and non-Rails as well.

**Comment by jdantonio** [comment]
@pitr-ch We haven't heard from you in a while? Are you still interested in maintaining this gem? I haven't worked with Ruby at all in the past two years and don't expect that to change in the foreseeable future. But many people depend on this gem so we need to make sure it's well maintained.

**Comment by jrochkind** [comment]
Uh oh, it's a dependency of Rails now too, it def needs to be well-maintained. It's whole raison d etre IMO is being well-maintained standard shareable concurrency primitives. 

I would not assume not commenting on _this_ issue (which is Rails-specific, and concurrent-ruby is not) means a lack of maintenance. But if there is a lack of maintenance, it would be a problem to many people's plans, and should be made transparent and dealt with. This thing existing is hugely useful to me and many others, but only if it's maintained.

**Comment by jdantonio** [comment]
I stepped down as maintainer in October of last year (issue #580), but I still follow notifications. If @pitr-ch isn't interested in maintaining the gem anymore I'll step up my involvement for a bit to help with these outstanding issues. But the best long-term solution is for someone who works regularly in Ruby to take the lead.

**Comment by matthewd** [comment]
@jrochkind right, that's exactly what I wanted to do: allows Rails (or any other interested library) to automatically wrap futures when it is necessary to ensure [high-concurrency] safety.

The cost should be minimal (and if it's not, we can and should fix that in Rails)... and should only be paid if it's needed, in which case the cost is arguably immaterial anyway.

The fact @edwardmp is seeing runtime autoloading at all would imply that `eager_load` is disabled, so that should settle your disturbance there.

As I mentioned in March, I really wanted/want to address this transparently in Rails so it'll Just Work for everyone. But @pitr-ch has made it clear he's not prepared to provide the API we need to do that, because some other library might misuse it; and I'm really not keen on the idea of monkey-patching c-r to achieve it.

**Comment by jrochkind** [comment]
Thanks @matthewd. I'm confused, is @pitr-ch a Rails committer too?

What were your thoughts/plans for how to do this transparently?  Whether it's part of rails, part of c-r, an extra gem, a monkey-patch, whatever -- you're the expert here, and I think the fact that this stuff doens't Just Work in a Rails environment is a real barrier to increasing use of concurrency in places it makes sense in Rails-land, and I'd like to figure out what's up and see if we can provide some solutions. 

It does not seem totally appropriate to have c-r itself referring to Rails classes like `Rails.application.executor.wrap` -- I doubt that was your thought. What was your thought? If you wanted to share an outline, it would be helpful to many of us I think.

**Comment by jdantonio** [comment]
I'm not familiar with the issue or its history so I'll have to catch up before I can express an opinion. I'm probably less concerned about people misusing a feature than @pitr-ch is. This is a concurrency library after all. Proper use of the more advanced featured requires a certain level of knowledge to begin with. I don't see why we can't work something out.

**Comment by chrisseaton** [comment]
I can spend a couple of hours a week trying to get the issue count down in general.

**Comment by matthewd** [comment]
My plan involves c-r providing a low-level API (which Rails would call), to say essentially "please wrap every future-execution with this block" (the executor), and possibly another to wrap the value-wait (the permit). It's not elegant, but it's the only Works By Default option that I can see -- I'm certainly open to alternative suggestions, though.

---

[The first time I brought it up](https://gitter.im/ruby-concurrency/concurrent-ruby?at=581719257b15d16e55c9202c), @pitr-ch threatened to remove the default executors ("This discussion helped me realise [the current API is too convenient]"), and made it clear he was using his maintainership of this library to "keep advocating against" (potentially questionable, but very present in the language) core ruby design concerns.

[The second time](https://gitter.im/ruby-concurrency/concurrent-ruby?at=586ec06a9d4cc4fc537e7ebe), he implied I (or, my requested API) was disrespecting / not "appreciating" c-r's users, and said he would not knowingly support a situation where c-r provided an API that two libraries could use in a conflicting manner, because it might happen and that's a good enough reason to disallow it.

It was about there that I [gave up](https://github.com/ruby-concurrency/concurrent-ruby/issues/585#issuecomment-290266906).

**Comment by chrisseaton** [comment]
@matthewd knowing @pitr-ch well I'm sure he's trying to do the right thing and is open to ideas and compromises and any bad feeling here is accidental. Please don't give up communicating what you need. I'll try to make time to read back through this issue and see what the issues are as well.

**Comment by jrochkind** [comment]
I think it would be worth revisiting with more maintainers/committers involved. 

But if it can't/won't be made part of c-r, the options are:

1. monkey-patching c-r
2. providing 'wrapper' classes in another gem that proxy to c-r. 

I think it's worth considering either, if neccesary. We need a way to do concurrency in rails that isn't a mess. Sadly/ironically/paradoxically, @matthewd's improvements to Rails 5.1 which make it _possible_ to do auto-loading in a reliably concurrency-safe way for the first time -- also mean you _need_ to use Rails API when doing any concurrency, or you get a deadlock (if there's any auto-loading going on).  Which is a big problem, and makes it a heck of a lot harder to do concurrrency. Before, you'd get occasional race conditions you could just ignore if you had auto-loading on--now you get reproducible deadlocks instead, if you don't use Rails api. 

While I understand and appreciate the intent, I'm not super happy with this outcome. Even if we fix c-r to work somehow, there are libraries and people doing multi-threaded concurrency without c-r too. But anyways. 

Ideally there should be a way not only to easily write concurrent stuff for Rails, but a way to use concurrent stuff written not neccesarily with rails in mind safely in rails as well. With regard to c-r, that probably means a monkey-patch rather than a proxy class... I think probably. I'm willing to consider it as a developer-user.

**Comment by jrochkind** [comment]
@matthewd btw, something that would still really help is still docs... the docs in this 10-month-old un-merged PR are actually super helpful, it would be nice if they were more discoverable than they are in an un-merged PR. https://github.com/rails/rails/pull/27494

Anyone on this thread trying to figure out what's we're talking about in Rails, I think the docs in that un-merged PR are probably the best/only overview.

**Comment by jdantonio** [comment]
All: I emailed @pitr-ch on his private email. He says he wants to be the maintainer of this gem, admits he hasn't been very responsive, and has promised to talk to me more over the next few days. I've worked with him on this gem for several years and I know he cares very deeply about its success. Please give us a chance to talk things over and figure out what's best for everyone. Unless there is an immediate need, can we revisit this after the American Thanksgiving holiday?

**Comment by pitr-ch** [comment]
Hi sorry, I was very busy over the summer and I let myself ignore notifications from c-r. However I did not intend to misconfigure my filters for mentions, I thought nobody was looking for me personally. I'll catch up, and please accept my apologies.

**Comment by edwardmp** [comment]
@jrochkind 

Thanks for your response. My settings relating to autoload should be pretty default.
Indeed in production `cache_classes = true` and `eager_load = true`.

That being said, this issue happens pretty infrequently in my production environment, the last occurrence was about 50 days ago. I currently do not wrap my code using stuff like `Rails.application.executor.wrap` anymore.

CC @matthewd TLDR; oddly enough `eager_load` is not disabled, so that's not what's happening in my case, very strange.

**Comment by matthewd** [comment]
@edwardmp maybe you have entries in `autoload_paths` that aren't in `eager_load_paths`? :confused:

If everything is eager loaded, there shouldn't be anything left to autoload.

**Comment by jrochkind** [comment]
@matthewd this is kind of off-topic but it's something I _have not_ been able to figure out (and it is relevant to using rails with concurrency safely) -- is there any way to actually _turn off_ autoloading in Rails?  I have not been able to find it in any guides or docs, and have not been able to figure out the code. Like even if you have entries in autoload_paths that aren't in eager_load_paths, just say "no, don't do autoloading" -- cause I'd rather get a clear "constant not found" error than a deadlock. 

(Also makes one think it should not be _allowed_ to have things in autoload_paths that aren't in eager_load_paths, rails should not allow it. The consequences of doing something weird here get more drastic in rails 5.1, and this has always been a confusing high-churn part of Rails).

**Comment by matthewd** [comment]
[Yes](https://github.com/rails/rails/blob/56c1326abb11ed275f04b6e0592ca66975e37f24/railties/lib/rails/application/finisher.rb#L190-L191): http://guides.rubyonrails.org/upgrading_ruby_on_rails.html#autoloading-is-disabled-after-booting-in-the-production-environment

Also part of my confusion about how @edwardmp is getting far enough into the autoloader to see the "circular dependency detected" message.

**Comment by pitr-ch** [comment]
Hi @matthewd, sorry I believe there was a misunderstanding before you gave up, and I certainly did not help with my lack of communication. Lets continue working on the solution. My last thinking was to introduce a private undocumented API for Rails to hook into to fix this. This API although undocumented would be tested and maintained for Rails. However I would like to keep it hidden because I don't want other projects/libraries to use it. If possible Rails should eventually stop using it as well.

@matthewd would this solution work for Rails? If you think so I'll introduce the hidden API to c-r pool in a PR and I'll add a prototype of a Rails fix. @matthewd would you be so kind to review and finish the prototype?

**Comment by edwardmp** [comment]
@matthewd 
`
    extra_dirs = ["#{Rails.root}/app/models/validators", "#{Rails.root}/app/jobs/concerns"]
    config.autoload_paths += extra_dirs
    config.eager_load_paths += extra_dirs
`
This is my config. So that can't be it. Great to see some activity in this thread though!

**Comment by jrochkind** [comment]
I really don't like the idea of a private undocumented API for Rails only. 

1) Why should Rails get special access?

2) It's pretty hard to figure out what's going on in Rails-land with regard to concurrency and autoloading, by either docs or code reading -- using private undocumented API here will not make it any easier. 

@pitr-ch I agree this is not ideal -- I think we can basically just blame autoloading, which has been hard to get right through the entire history of Rails (and isn't concurrency-safe even for ruby stdlib autoload I think), but is a much-loved feature anyhow. This non-ideal situation comes about from @matthewd's heroic efforts to make auto-loading thread-safe -- but only if you do special Rails things in all your threads, alas.   But it might help if you elucidated the ways you think this could be abused by other code -- just hooking something in there that the main app developer was unaware of, doing odd things?  

@matthewd I'm trying to think of other ways to make this work without a gotcha in rails land. I think, maybe it would make sense if you could _turn off_ autoloading in these non-properly-rails-done blocks.  But since these 'dangerous' blocks are by definition in other threads, that would still require a custom wrapper of `without_autoloading do` in all your threads. 

Hmm, what if we approach from the other direction: What if autoloading were only ON inside proper `executor.wrap` blocks? Rails just wouldn't even try to autoload at all unless it had the proper lock. Because it's not safe to do so unless you have the proper lock. So you have two choices, use the proper Rails API, or don't and you won't get any auto-loading, but won't get deadlocks either.  For backwards compatibility purposes, perhaps this is an opt-in configuration (although it's _wrong_ to autoload without the lock, perhaps people have found ways to make it work anyway and wouldn't want to break them).  I'm not totally sure this would work -- if no autoloading happened in the thread (that didn't use `executor.wrap` so no locks were taken), what are the implications on presence or absence of `interlock.permit_concurrent_loads`?  

But if it would actually work, I think it may be the ideal solution, that doesn't require messing with every use of concurrency everywhere in a Rails app.  Autoloading is nice, but it just doesn't play well with concurrency. @matthewd has provided a genius opt-in way to make it work well with concurrency -- but could that now present architecture also be used to just disable auto-loading if you haven't opted in?

**Comment by matthewd** [comment]
@edwardmp  sounds like an issue for rails/rails if you see it again, then. 🤷🏻‍♂️

@pitr-ch Rails will never stop using it, and as long as that's a hoped-for goal, I'm nervous it will go away. (Ruby constant lookup is independent of threads; if that were to change, I'm not sure the result would still be Ruby. And as long as multiple threads are looking up constants in the same place, we're going to need a way to 1. stop the world while we're modifying it\*, and 2. defer modifying it when that could interfere with other threads\*\*.)

I'd much prefer something that was documented, with a warning that it's dangerous and has risks similar to monkey-patching Ruby core classes, and is only suitable for careful use by end applications and full-coverage frameworks -- or something along those lines. A small doc page on 'framework support' covering what it allows and why it's a bad idea, then a link from the relevant couple of methods, say.

I'd rather not have something that is explicitly "just for Rails", both because that effectively deprives Rails-competitors of equal footing, and it means you miss out on an opportunity to make your case *against* people using it. They'll still find it used in Rails, and if it's undocumented, ISTM they're far more likely to just copy our usage without understanding the implications.

Frankly, in a language that freely allows a developer to override String#+ and then just trusts they won't (unless they feel they have a good reason), protecting users from other libraries' possible misdeeds just seems impossible to the degree of pointlessness. I'm happy to argue usefulness of possible APIs vs their implementation trade-offs all day, but for a library that wants to position itself as "stdlib 2.0" to take such an anti-Ruby design stance leaves me feeling stuck. If I wanted to work in a language whose designers and library authors restricted my actions, not for technical reasons, but because they knew better than me, I'd be using [insert other language here]. 😐

As for the technical side: I'll emphasize that I need to wrap *all* execution, not just a single executor -- that includes the global ones, but also any others that people create. Does this mean I could break all sorts of things if I'm not very careful? Yes, yes it does. But that's my / the framework's job: we take control of many core behaviours in order to give people a consistent experience. And when we do screw that up, I think history shows that people have no difficulty working out who to blame. :wink:

---

\* this one could actually be addressed by a more flexible `autoload` implementation in MRI. But it seems that's never going to happen. :cry:

\*\* no getting around this one, though.

**Comment by edwardmp** [comment]
@matthewd there's already a Rails issue I created a few months ago (linked earlier in the thread): https://github.com/rails/rails/issues/30350. If you have any further ideas indeed it's best to discuss there, to keep this thread clean :).

**Comment by matthewd** [comment]
> we can basically just blame autoloading, which has been hard to get right through the entire history of Rails (and isn't concurrency-safe even for ruby stdlib autoload I think)

This is the great irony: we have to limp along with `const_missing`, and thus do our own locking to provide some thread safety (now with the interlock, previously by wrapping a mutex around the whole application), whereas Ruby's internal `autoload` *is* threadsafe (by design, at least -- I've seen some indications it's not always so perfect in practice). But because `autoload` is scheduled for the chopping-block, no-one's willing to add new features that we'd need to be able to use it for our purposes. (Mostly, we need to autoload blocks, instead of files... otherwise we'd be creating a *lot* of tempfiles.)

Don't even get me started on Digest's non-thread-safe use of `const_missing`, and the hack of a workaround it recommends, just because there's a cloud over `autoload`.

---

@jrochkind we're supposed to have that use case covered by the (**still** undocumented 🙄) `config.allow_concurrency = :unsafe` setting, which was intended to behave approximately the same way that 4.2 did in production. (That one was intentionally undocumented though -- it's there so I can point people to it if they describe some usage that previously worked fine but is somehow fundamentally incompatible with the interlock. With the extended thread use of the 5.0 era, though, it's even less of a good idea.)

I have indeed been thinking about having autoloading just raise if it's not inside an executor block... it technically doesn't solve the problem, because we'd be detecting the wrong thing: the deadlock is because the *other* (waiting) thread isn't permitting the load. But forcing people to address one might have a better chance that they'd get the other right too.

Correspondingly, the risk that non-executor code is exposed to is not from an autoload in that thread (which we could control). The problem is when that thread concurrently accesses a constant which a different (properly behaving, lock-respecting) thread is in the middle of loading that same constant: our unprotected thread can end up getting, and using, a class that doesn't have any methods yet. (Or far worse, if it outlives the request and a reload occurs.)

Overall, the whole point of the locking is that the threads need to cooperate in order to avoid interfering with *each other*. An "uncontrolled" thread hitting an autoload is the best-case scenario: it's just as likely to hit a totally 'unexplainable' error, because basic "how ruby works" expectations break down.

**Comment by jrochkind** [comment]
@matthewd my understanding of `config.allow_concurrency = :unsafe ` is that it turns off _concurrent request handling_.  I do not want to do that. 

I want concurrent request handling to happen in production. I want auto-loading to happen in dev. 

I want _non-compliant threads_ (which are NOT Rails requests threads, which are perfectly compliant with the new APIs, but are likely to be little tiny things doing things like making 5 HTTP fetches concurrently) to _not autoload_.  

This may be unreasonable for reasons beyond me? But I don't believe it's what `config.allow_concurrency = :unsafe` does, at all.

**Comment by pitr-ch** [comment]
Wrapping all c-r tasks with `Rails.application.reloader.wrap` is doable, however wrapping each blocking method like `#value` with `permit_concurrent_loads` is not. There is too many of them, it would be a nightmare to maintain, keeping it correct. 

Based on my limited understanding so far, it should be enough that all c-r tasks are wrapped in `Rails.application.reloader.wrap` (they can do a Rails work) and `ActiveSupport::Dependencies.interlock.permit_concurrent_loads` (they can trigger concurrent auto load). Then `#value` should not lead to any problems, since it reads already completed value or it will potentially triggers execution of already wrapped task on a pool. (I need to check there is no abstraction which would execute task itself without wrapper.) @matthewd Please correct me if I am wrong.

**Comment by pitr-ch** [comment]
@jrochkind the rationale against global state: concurrent-ruby is on a direction to get rid off all global state like configuration and global pools. If a library is designed with a global state, eventually a conflict will arise where 2 different users (projects, other libraries) of the global state will require different values there. E.g. conflicts in: configuration of concurrent-ruby's exception handling (an example there is actually no such configuration), configuration of global pools, non-compostability of task wrappers which is the problem of the API discussed here. If there is multiple users (libs, projects, frameworks) of task wrapper API, the users cannot know how to compose the wrappers together, how they depend in each other, if they are compatible.

Side-note: Global state is bad in libraries but of course a developer (having an knowledge about the whole project, dependencies, and source code) can safely decide to put a global pool into a constant and use it it in the whole project with easy accesses. Library never has a global view of the whole project and therefore it should not have global state.

@matthewd I assume an agreement that once introduced (even if as "secret" api) it will not go away until Rails no longer need it.

I should have written publicly undocumented. I assumed it would have documentation and explanation in the source code only. So a special documented private API which would not appear in the generated documentation. (This does not apply any more, see below.)

With the upcoming deprecation of global state there will be alternative to this API, then Rails can switch to the new API. This is what I hope for not that the need for the wrapper will go away.

Yes, Ruby can be messy sometimes. If I can, I don't want to add to that messiness though.

**TL;DR:** After thinking about it more I'll add an API to wrap all asynchronously executed tasks, including public documentation with proper warnings. 

Reasoning for the record: 
* The wrapping API is something which is obviously needed for async executions, it should have been added from the beginning. 
* We introduced global state (and pools) which was a mistake, however until there is an alternative fixes and evolution of the old approach should not be blocked by a vision of cleaner non-global approach. 
* The wrapping API is essential for the global pools in c-r, it has to be added even if it's a bad API. In the end it will be deprecated along with all the global pools, so it's fine. (There will be a replacement for all the deprecated features of global pools none of the features will be lost.)
* I failed to look at the global pools in isolation apart from future direction of complete global-state removal.

**Technical part:** Yes, of course, any asynchronously executed task in c-r created by any of the abstractions has to be wrapped.

I'll make sure that all async execution is wrapped when the a wrapper is set in the API.

Why would `#value` method need to be wrapped in `permit_concurrent_loads`? If all async execution is already properly wrapped then `#value` just reads already resolved value or triggers execution of an async task which is already wrapped.

**Comment by matthewd** [comment]
I disagree that the convenience pools were/are a mistake -- I think a big part of the appeal of this library is that you can fire-and-forget tasks when you don't want to manage a pool, etc, and instead just know that they'll happen. If you become less convenient than `Thread.new { ... }.join` just for the sake of purity, I think people will follow their fingers. But that's a separate API consideration from what we're focused on here, so I won't dwell on it.

My confusion is that you keep bringing up the global pools as if they're to blame for this situation, but they're not: I do not want to wrap everything that the global pool does. I want to wrap everything that *every* executor does. (Similarly, removing the global pools does not solve that: I will still need to wrap *everything*.)

---

`executor.wrap` acquires a lock (among other things); `permit_concurrent_loads` temporarily releases that lock. So it has to go immediately around the blocking operation: a thread inside `permit_concurrent_loads` is very restricted in what it can safely do.

This one matters less, though, because the more I think about it, the more I think I should just monkey-patch `Mutex#lock` and `ConditionVariable#wait` directly.

**Comment by jrochkind** [comment]
I also disagree that the global pool are a mistake. I would recommend you leave them.

**Comment by jrochkind** [comment]
@pitr-ch If I understand what's going on with `permit_concurrent_loads` in my examples above, in some cases the threads are actually _blocked and not processing_ until I wait for them on `value` with `permit_concurrent_loads`.  That `permit_concurrent_loads` -- causing the waiting thread to temporarily release it's lock --  is part of what made them actually run-able.  But I'm not worried about it, because I think it only happens in certain circumstances -- when a thread is _actually_ triggering auto-loading, which only generally happens on the first same-bodied thread to be executed, and never in production. 

I also could be misunderstanding what's going on.

**Comment by pitr-ch** [comment]
Let's move the discussion about global state and pools removal to https://github.com/ruby-concurrency/concurrent-ruby/issues/595. I'll later reply there later and I hope I'll dismiss your concerns.

**Comment by pitr-ch** [comment]
Some more details about the planned pools removal are in https://github.com/ruby-concurrency/concurrent-ruby/issues/595#issuecomment-346490129

Future impact on Rails and other global state embracing libraries/frameworks: they will no longer be able to reuse c-r global pools. The libraries will be required to create their global world and document it that its the one correctly configured (and with the wrappers) to be used by users of the given library. If users combine libraries/projects/frameworks using c-r together it's expected that they will configure one world if possible and inject into those libraries. However the big advantage is that they will have full controls over the configuration and tools how to deal with conflicts. Rather than in current situation where you might need to settle for compromising configuration, or 2 ruby processes to isolate.

@jrochkind thanks for the additional details. However that's quite bad. Wrapping every `#value` like call, which implies really any blocking call in c-r, does not sound feasible. I am hoping there is another solution to be discovered.

@matthewd `Mutex#lock` and `ConditionVariable#wait` patching does not sound good as well. All blocking calls would have to be patched. Any new blocking calls introduces in a library (cext) not using any of the previously patched methods would not work.

I'll need to take some time to understand the Rails autoloading locking in detail.

**Comment by matthewd** [comment]
> [Rails] will be required to create their global world and document it that its the one correctly configured (and with the wrappers) to be used by users of the given library.

Yes, that was my earlier understanding, and exactly what I stated is not sufficient for our needs.

> I'll need to take some time to understand the Rails autoloading locking in detail.

Don't bother. If the future plan is that Rails will need to supply its own concurrency API (regardless of whether it's internally built using c-r), and anyone not using the Rails-specific one will be out of our control, then I might as well go and do that now.

**Comment by jrochkind** [comment]
I think removing the global pools is a mistake, but that's a different topic. I frequently use c-r in places where I'm just using a sprinkling, and the ease of setup to just create a future and use it, without having to create my own infrastructure, is huge.  If you have documentation somewhere I can read that would make the case for why the global pools are bad, I'd be interested in reading it.

**Comment by mrbongiolo** [comment]
So, today I've been bitten by those deadlocks, it wasn't nice!
The worst thing is that there is no exception, nothing, just the server hanging there. I was trying a lot of things, and then, eventually I came across a "Circular reference exception" is some of my specs (but it just happened here and there), which finally gave me some hints to find this issuer right here.

After reading over it I understood that to make r-c work with Rails, I should wrap most of the calls around that `executor.wrap` and also use `AS::D.interlock.permit_concurrent_loads`, it seems to be working now.

My main issue, in this case, was these deadlocks without a single hint on where to start debugging, the guides on rails are scarce in this subject and not so easy to find, I found them linked from this issue. 

It would be really helpful to have a small section in r-c README pointing out to these docs (https://guides.rubyonrails.org/threading_and_code_execution.html#executor) when someone wants to use it with Rails 5, since this is the main "to-go" lib when someone wants to add concurrency, and using it with Rails is probably a common practice, so it's easier to be aware of what is needed to make it work in Rails from a README here.

If you want, I can event open a PR with a text to add to it.

And thanks for this awesome lib!!!

**Comment by luizkowalski** [comment]
Hey! not sure if this is the best place to do it but I stumbled upon this issue a couple of weeks ago when adding Futures to my application. after a lot of back and forth with the code, trials and errors and dig through documentations, I created a small library to help me use Futures on Rails applications:

https://github.com/luizkowalski/concurrent_rails

It is simply a wrapper for Future while accounting for Rails executor. 

Thought you guys would like to know even though this issue is stale for quite some time now :)

**Comment by joelvh** [comment]
@luizkowalski I just came across this thread and was wondering when you'll publish your gem?

**Comment by luizkowalski** [comment]
@joelvh hey, just did! forgot to do it before because I had the first version yanked due to bad configuration and then I forgot

**Comment by Hungor** [comment]
@jrochkind Thanks for the reply! You just saved my day!
